File: client/src/main/scala/org/zalando/nakadi/client/ClientImpl.scala
package org.zalando.nakadi.client

import scala.Left
import scala.Right
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.concurrent.duration.DurationInt
import org.slf4j.LoggerFactory
import org.zalando.nakadi.client.model._
import com.typesafe.scalalogging.Logger
import Client.ClientError
import akka.actor.Terminated
import akka.http.scaladsl.model.HttpHeader
import akka.http.scaladsl.model.HttpResponse
import akka.http.scaladsl.model.HttpResponse
import akka.http.scaladsl.model.StatusCodes
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.http.scaladsl.model.headers.RawHeader

private[client] class ClientImpl(connection: Connection, charSet: String = "UTF-8") extends Client with HttpFactory{
  import Client._
  implicit val materializer = connection.materializer

  val logger = Logger(LoggerFactory.getLogger(this.getClass))
  def metrics()(implicit ser: NakadiDeserializer[Metrics]): Future[Either[ClientError, Option[Metrics]]] = {
    logFutureEither(connection.get(URI_METRICS).flatMap(mapToEither(_)))
  }

  def eventTypes()(implicit ser: NakadiDeserializer[Seq[EventType]]): Future[Either[ClientError, Option[Seq[EventType]]]] = {
    logFutureEither(connection.get(URI_EVENT_TYPES).flatMap(mapToEither(_)))
  }

  def newEventType(eventType: EventType)(implicit ser: NakadiSerializer[EventType]): Future[Option[ClientError]] = {
    logFutureOption(connection.post(URI_EVENT_TYPES, eventType).flatMap(mapToOption(_)))
  }

  def eventType(name: String)(implicit ser: NakadiDeserializer[EventType]): Future[Either[ClientError, Option[EventType]]] = {
    logFutureEither(connection.get(URI_EVENT_TYPE_BY_NAME.format(name)).flatMap(in => mapToEither(in)))
  }

  def updateEventType(name: String, eventType: EventType)(implicit ser: NakadiSerializer[EventType]): Future[Option[ClientError]] = {
    val result = connection.put(URI_EVENT_TYPE_BY_NAME.format(name), eventType)
    logFutureOption(result.flatMap(in => mapToOption(in)))
  }

  def deleteEventType(name: String): Future[Option[ClientError]] = {
    logFutureOption(connection.delete(URI_EVENT_TYPE_BY_NAME.format(name)).flatMap(in => mapToOption(in)))
  }

  def newEvents[T](name: String, events: Seq[T])(implicit ser: NakadiSerializer[Seq[T]]): Future[Option[ClientError]] = {
    logFutureOption(connection.post(URI_EVENTS_OF_EVENT_TYPE.format(name), events).flatMap(in => mapToOption(in)))
  }

  def events[T](name: String, params: Option[StreamParameters])(implicit ser: NakadiDeserializer[T]): Future[Either[ClientError, Option[T]]] = {
    val headers = withHeaders(params)
    logFutureEither(connection.get(URI_EVENTS_OF_EVENT_TYPE.format(name)).flatMap(in => mapToEither(in)))
  }

  def partitions(name: String)(implicit ser: NakadiDeserializer[Partition]): Future[Either[ClientError, Option[Partition]]] = {
    logFutureEither(connection.get(URI_PARTITIONS_BY_EVENT_TYPE.format(name)).flatMap(in => mapToEither(in)))
  }

  def partitionById(name: String, id: String)(implicit ser: NakadiDeserializer[Partition]): Future[Either[ClientError, Option[Partition]]] = {
    logFutureEither(connection.get(URI_PARTITION_BY_EVENT_TYPE_AND_ID.format(name)).flatMap(in => mapToEither(in)))
  }

  def validationStrategies()(implicit des: NakadiDeserializer[Seq[EventValidationStrategy.Value]]): Future[Either[ClientError, Option[Seq[EventValidationStrategy.Value]]]] = {
    logFutureEither(connection.get(URI_VALIDATION_STRATEGIES).flatMap(mapToEither(_)))
  }

    def enrichmentStrategies()(implicit des: NakadiDeserializer[Seq[EventEnrichmentStrategy.Value]]): Future[Either[ClientError, Option[Seq[EventEnrichmentStrategy.Value]]]] = {
    logFutureEither(connection.get(URI_ENRICHMENT_STRATEGIES).flatMap(mapToEither(_)))
  }

  def partitionStrategies()(implicit des: NakadiDeserializer[Seq[PartitionStrategy.Value]]): Future[Either[ClientError, Option[Seq[PartitionStrategy.Value]]]] =
    logFutureEither(connection.get(URI_PARTITIONING_STRATEGIES).flatMap(mapToEither(_)))

  def stop(): Future[Terminated] = connection.stop()

  //####################
  //#  HELPER METHODS  #
  //####################

  private[client] def logFutureEither[A, T](future: Future[Either[ClientError, T]]): Future[Either[ClientError, T]] = {
    future recover {
      case e: Throwable =>
        logger.error("A unexpected error occured", e)
        Left(ClientError("Error: " + e.getMessage, None))
    }
  }
  private[client] def logFutureOption(future: Future[Option[ClientError]]): Future[Option[ClientError]] = {
    future recover {
      case e: Throwable =>
        logger.error("A unexpected error occured", e)
        Option(ClientError("Error: " + e.getMessage, None))
    }
  }

  private[client] def mapToEither[T](response: HttpResponse)(implicit deserializer: NakadiDeserializer[T]): Future[Either[ClientError, Option[T]]] = {
    logger.debug("received [response={}]", response)
    response match {
      case HttpResponse(status, headers, entity, protocol) if (status.isSuccess()) =>
        try {
          Unmarshal(entity).to[String].map(body => Right(Some(deserializer.fromJson(body))))
        } catch {
          case e: Throwable =>
            val msg = "Failed to deserialise the content with error: %s".format(e.getMessage)
            logger.error(msg)
            Future.successful(Left(ClientError(msg, Some(status.intValue()))))
        }
      case HttpResponse(StatusCodes.NotFound, headers, entity, protocol) =>
        Future.successful(Right(None))
      case HttpResponse(status, headers, entity, protocol) if (status.isRedirection()) =>
        val msg = "Not implemented: http-status (" + status.intValue() + "}) and reason:" + status.reason()
        logger.info(msg)
        Future.successful(Left(ClientError(msg, Some(status.intValue()))))
      case HttpResponse(status, headers, entity, protocol) if (status.isFailure()) =>
        Unmarshal(entity).to[String].map { body =>
          val msg = "An error occurred, http-status: %s (%s) Message: %s".format(status.intValue(), status.reason(), body)
          logger.warn(msg)
          Left(ClientError(msg, Some(status.intValue())))
        }
    }
  }

  private[client] def mapToOption[T](response: HttpResponse): Future[Option[ClientError]] = {
    response.status match {
      case status if (status.isSuccess()) =>
        logger.debug("Success. http-status: %s", status.intValue().toString())
        Future.successful(None)
      case status if (status.isRedirection()) =>
        val msg = "Redirection - http-status: %s, reason[%s]".format(status.intValue().toString(), status.reason())
        logger.info(msg)
        response.entity.toStrict(10.second).map { body =>
          logger.debug("Redirection - http-status: %s, reason[%s], body:[%s]".format(status.intValue().toString(), status.reason(), body.data.decodeString(charSet)))
        }
        Future.successful(Option(ClientError(msg, Some(status.intValue()))))
      case status if (status.isFailure()) =>
        response.entity.toStrict(10.second).map { body =>
          val msg = "Failure - http-status: %s, reason[%s], body:[%s]".format(status.intValue().toString(), status.reason(), body.data.decodeString(charSet))
          logger.warn(msg)
          Option(ClientError(msg, Some(status.intValue())))
        }
    }
  }

}

File: client/src/main/java/org/zalando/nakadi/client/java/ClientImpl.java
package org.zalando.nakadi.client.java;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.Future;

import org.zalando.nakadi.client.Deserializer;
import org.zalando.nakadi.client.Serializer;
import org.zalando.nakadi.client.java.enumerator.EventEnrichmentStrategy;
import org.zalando.nakadi.client.java.enumerator.PartitionStrategy;
import org.zalando.nakadi.client.java.model.Event;
import org.zalando.nakadi.client.java.model.EventStreamBatch;
import org.zalando.nakadi.client.java.model.EventType;
import org.zalando.nakadi.client.java.model.Metrics;
import org.zalando.nakadi.client.java.model.Partition;
import org.zalando.nakadi.client.java.utils.SerializationUtils;
import org.zalando.nakadi.client.utils.Uri;
import com.fasterxml.jackson.core.type.TypeReference;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;

public class ClientImpl implements Client {
    private final JavaClientHandler handler;

    // Deserializers
    private final Deserializer<Metrics> metricsDeserializer = SerializationUtils.metricsDeserializer();
    // List Deserializers
    private final Deserializer<List<EventType>> seqOfEventTypeDeserializer = SerializationUtils.seqOfEventTypeDeserializer();
    private final Deserializer<List<Partition>> seqOfPartitionDeserializer = SerializationUtils.seqOfPartitionDeserializer();
    private final Deserializer<List<EventEnrichmentStrategy>> seqOfEventEnrichmentStrategy = SerializationUtils.seqOfEventEnrichmentStrategy();
    private final Deserializer<List<PartitionStrategy>> seqOfPartitionStrategy = SerializationUtils.seqOfPartitionStrategy();
    // Serializers
    private final Serializer<EventType> eventTypeSerializer = SerializationUtils.defaultSerializer();
    private final Deserializer<EventType> eventTypeDeserializer = SerializationUtils.eventTypeDeserializer();

    public ClientImpl(JavaClientHandler handler) {
        this.handler = handler;
    }

    @Override
    public Future<Optional<Metrics>> getMetrics() {
        return handler.get(Uri.URI_METRICS(), metricsDeserializer);
    }

    @Override
    public Future<Optional<List<EventType>>> getEventTypes() {
        return handler.get(Uri.URI_EVENT_TYPES(), seqOfEventTypeDeserializer);
    }

    @Override
    public Future<Void> createEventType(EventType eventType) {
    	checkNotNull(eventType, "EventType may not be null");
        return handler.post(Uri.URI_EVENT_TYPES(), eventType, eventTypeSerializer);
    }

    @Override
    public Future<Optional<EventType>> getEventType(String eventTypeName) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
        return handler.get(Uri.getEventTypeByName(eventTypeName), eventTypeDeserializer);
    }

    @Override
    public Future<Void> updateEventType(String eventTypeName, EventType eventType) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
    	checkNotNull(eventType, "EventType may not be null");
       return handler.put(Uri.getEventTypeByName(eventTypeName), eventType, eventTypeSerializer);
    }

    @Override
    public Future<Void> deleteEventType(String eventTypeName) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
        return handler.delete(Uri.getEventTypeByName(eventTypeName));
    }

    @Override
    public <T extends Event> Future<Void> publishEvent(String eventTypeName, T event, Serializer<List<T>> serializer) {
    	checkNotNull(event, "Event may not be null");
        return publishEvents(eventTypeName, Arrays.asList(event), serializer);
    }

    @Override
    public <T extends Event> Future<Void> publishEvent(String eventTypeName, T event) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
    	checkNotNull(event, "Event may not be null");
        return publishEvents(eventTypeName, Arrays.asList(event));
    }

    @Override
    public <T extends Event> Future<Void> publishEvents(String eventTypeName, List<T> events,
            Serializer<List<T>> serializer) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
    	checkListIsNotEmptyOrNull(events, "List of events must not be null or empty");
        return handler.post(Uri.getEventStreamingUri(eventTypeName), events, serializer);
    }

    @Override
    public <T extends Event> Future<Void> publishEvents(String eventTypeName, List<T> events) {
    	checkListIsNotEmptyOrNull(events, "List of events must not be null or empty");
        return publishEvents(eventTypeName, events, SerializationUtils.defaultSerializer());
    }

    @Override
    public Future<Optional<List<Partition>>> getPartitions(String eventTypeName) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
        return handler.get(Uri.getPartitions(eventTypeName), seqOfPartitionDeserializer);
    }

    @Override
    public Future<Optional<List<EventEnrichmentStrategy>>> getEnrichmentStrategies() {
        return handler.get(Uri.URI_ENRICHMENT_STRATEGIES(), seqOfEventEnrichmentStrategy);
    }

    @Override
    public Future<Optional<List<PartitionStrategy>>> getPartitioningStrategies() {

        return handler.get(Uri.URI_PARTITIONING_STRATEGIES(), seqOfPartitionStrategy);
    }

    @Override
    public void stop() {
        handler.stop();
    }

    @Override
    public <T extends Event> Optional<ClientError> subscribe(String eventTypeName, StreamParameters parameters,
            Listener<T> listener, Deserializer<EventStreamBatch<T>> deserializer) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
        return handler.subscribe(eventTypeName, Uri.getEventStreamingUri(eventTypeName), parameters, listener,
                deserializer);
    }

    @Override
    public <T extends Event> Optional<ClientError> subscribe(String eventTypeName, StreamParameters parameters,
            Listener<T> listener, TypeReference<EventStreamBatch<T>> typeRef) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
    	checkNotNull(listener, "Listener may not be null");
        return handler.subscribe(eventTypeName, Uri.getEventStreamingUri(eventTypeName), parameters, listener,
                SerializationUtils.withCustomDeserializer(typeRef));
    }

    @Override
    public <T extends Event> void unsubscribe(String eventTypeName, Optional<String> partition, Listener<T> listener) {
    	checkStringIsNotNullOrEmpty(eventTypeName,"eventTypeName may not be null or empty");
    	checkNotNull(listener, "Listener may not be null");
        handler.unsubscribe(eventTypeName, partition, listener);
    }
    
    private <T> void checkListIsNotEmptyOrNull(List<T> l, String message){
    	Preconditions.checkArgument(l!=null&&!l.isEmpty(), message);
    	
    }
    private void checkStringIsNotNullOrEmpty(String in, String message){
    	Preconditions.checkArgument(!Strings.isNullOrEmpty(in), message);
    }
    private <T>void checkNotNull(T obj, String message){
    	Preconditions.checkNotNull(obj,message);
    }

}